### ISSUE TITLE
[API Proposal]: Create System.Text.Parsing

### ISSUE THREAD

**Issue by Foxtrek64** [dotnet/runtime#83652]
### Background and motivation

The System.Text.Parsing namespace will hold language-agnostic interfaces and classes used for parsing text-based languages such as Json, Xml, Yaml, and others. This could potentially be used by Roslyn as well, should they find anything useful, but that's not a priority of this library.

Additionally, this library will not be functional on its own. It should merely provide the framework on which to build language-specific parsing libraries, for example System.Text.Json or a future System.Text.Xml. Changes to System.Text.Json to use this library will be a project in itself and thus will be placed in its own proposal.

The S.T.Parsing library should begin its life by abstracting components of System.Text.Json into agnostic forms, wherever possible. This will give us a good head start, but additional features can be added to this library as we find a need for them, even if such features are not applicable to every possible language.

### API Proposal

First, a few more generic things that won't follow the standard suggestion format.

1. Migrate PooledByteBufferWriter.cs to System.Text.Parsing.PooledByteBufferWriter. Probably change visibility to `protected`, but I'm not too familiar with where this is used.
2. Migrate HexConverter.cs to System.Text.Parsing.HexConverter. Probably change visibility to `public`, else `protected`.
3. Migrate JsonNamingPolicy.cs to System.Text.Parsing.NamingPolicy. JsonNamingProperty should become a static class behaving like a type-safe enum, and specific policies like SnakeCaseNamingPolicy should instead implement NamingPolicy.

We should consider creating a bunch of generic Naming Policies. For instance, a SnakeCaseNamingPolicy could be applied to json as `"some_key": 42` and xml as `<some_key>42</some_key>`, so there is reuse to be had here.

Create System.Text.Parsing.Document.ITextDocument (or IDocument, whichever sounds better):
```cs
// I have only included public interfaces from JsonDocument here, but it may be prudent to expose some internal methods as well.
public interface ITextDocument<TRoot, TextWriter> : IDisposable
    where TRoot : IElement
    where TextWriter : ITextWriter
{
    void WriteTo(TextWriter writer);
}
```

Create System.Text.Parsing.Document.IElement
```cs
// Should these returns be generic somehow?
public interface IElement<TextWriter>
    where TextWriter : ITextWriter
{
    // Potentially some sort of generic ValueKind.
    
    // All elements should at least be readable. Consumer can make this writable if they choose.
    IElement this[int index] { get; }

    int GetArrayLength();

    IElement GetProperty(string propertyName);
    IElement GetProperty(ReadOnlySpan<char> propertyName);
    IElement GetProperty(ReadOnlySpan<byte> utf8PropertyName);

    bool TryGetProperty<TElement>(string propertyName, out TElement value) where TElement : IElement;
    bool TryGetProperty<TElement>(ReadOnlySpan<char> propertyName, out TElement value) where TElement : IElement;
    bool TryGetProperty<TElement>(ReadOnlySpan<byte> utf8PropertyName, out TElement value) where TElement : IElement;

    // Currently, there are separate GetInt(), GetByte(), etc. for the JsonElement type.
    X GetX();
    bool TryGetX(out x);

    // I propose we skip those and simply use generics. If the consumer wishes to separate these out, e.g. for backwards compatibility, that's fine, but with the new `INumber` interface to handle numeric types we shouldn't face any roadblocks with a generic implementation. And if something comes up where that is the case, the consumer can always explicitly implement these and have them throw, then implement separate GetX and TryGetX methods.
    TValue GetValue<TValue>();
    bool TryGetValue<TValue>(out TValue);

    string GetRawText();

    // Provided for backwards-compatibility
    bool ValueEquals(string? text);
    bool ValueEquals(ReadOnlySpan<byte> utf8Text);

    // Perform type comparison using default equality comparer.
    bool ValueEquals<TValue>(TValue other);

    void WriteTo(TextWriter writer);

    public ArrayEnumerator EnumerateArray();

    public ObjectEnumerator EnumerateObject();

    public override string ToString();

    public IElement Clone();
}
```

Create System.Text.Parsing.Document.JsonProperty
```cs
public interface IProperty<TextWriter>
    where TextWriter : ITextWriter
{
    string Name { get; }

    bool NameEquals(string? text);
    bool NameEquals(ReadOnlySpan<byte> utf8Text);
    bool NameEquals(ReadOnlySpan<char> text);

    bool WriteTo(TextWriter writer);

    string ToString();
}
```

Create System.Text.Parsing.Reader.IReader
Might also be a good candidate for an abstract class to implement a lot of the internal and private functionality.
```cs
public interface IReader
{
    ReadOnlySpan<byte> ValueSpan { get; private set; }

    long BytesConsumed { get; }

    long TokenStartIndex { get; private set; }

    int CurrentDepth { get; }

    // Some sort of generic TokenType

    bool HasValueSequence { get; private set; }
    bool ValueIsEscaped { get; private set; }
    bool IsFinalBlock { get; }
    
    ReadOnlySequence<byte> ValueSequence { get; private set; }

    SequencePosition Position { get; }

    ReaderState CurrentState { get; }

    bool Read();

    void Skip();
    bool TrySkip();

    bool ValueTextEquals(ReadOnlySpan<byte> utf8Text);
    bool ValueTextEquals(string? text);
    bool ValueTextEquals(ReadOnlySpan<char> text);
}
```

Create System.Text.Parsing.Writer.IWriter
Might also be a good candidate for an abstract base class.
I could go on forever I'm sure but this will get pretty repetitive. Discussion should be had for a list of candidates for migration/generalization.

### API Usage

This library contains base classes and interfaces and should only be used by those wanting to implement their own parsing/serialization library. No samples are provided here because it's just standard implementation.

### Risks

This proposal by itself should add no inherent risks as it is non-breaking. We are simply creating a new library and new types under its library.

**Comment by MihaZupan** [comment]
What would be the benefits of having such an abstraction?

**Comment by Foxtrek64** [comment]
@MihaZupan Primarily code deduplication. I also plan to propose System.Text.Xml as a replacement to System.Xml and it would require me to create exact copies of files available in S.T.J or copies with minor changes.

This will also be useful for people writing parsing libraries for other text config files or markup/down or whatever.

**Comment by MihaZupan** [comment]
The set of APIs you are proposing looks very biased towards a "text to POCO" data (de)serialization use case, not generic text parsing. Who would be the consumer of these abstractions -- when would someone use something like `IReader`?

> Primarily code deduplication

Can you give an example of code that could be deduplicated?

> This will also be useful for people writing parsing libraries for other text config files or markup/down or whatever.

How so? You mentioned markdown as an example, but I don't see what having a shared set of interfaces to implement would improve for a parsing library such as [Markdig](https://github.com/xoofx/markdig).

**Comment by Foxtrek64** [comment]
> The set of APIs you are proposing looks very biased towards a "text to POCO" data (de)serialization use case
This is correct. I'm perfectly fine changing the name of the proposed library if you feel something else is more appropriate.

> Who would be the consumer of these abstractions
Primarily those who wish to develop a library that allows for the (de)serialization of standardized text files. For instance, I recently had to write a library that would deserialize and serialize EDI X12 204 Motor Carrier Load Tender files, which is a text file for use with AS400 which represents an entire shipment history for a particular load. Having a library like System.Text.Parsing would have been very useful, but in its absense I had to roll my own tokenizer and (de)serializer.

> Can you give an example of code that would be deduplicated?
I can probably give a more precise example once I actually start looking at implementation, but as previously mentioned a lot of the naming policies would need to be duplicated across multiple code bases as well as the bodies of more generic things like documents, elements, and properties. I imagine a large majority of the code within JsonDocument for instance would be nearly identical to that of an XmlDocument, just swapping out references to certain things. For instance, Utf8JsonWriter would become Uft8XmlWriter, but both writers would share the exact same signatures a TextWriter.

> Markdown/Markdig
You have a good point here.  Not sure why I suggested Markdown honestly. This is too loose of a format to be (de)serialized in the same way as XML/Json/Yml/etc. Markup such as Xaml or Html could certainly be candidates, though that's cheating a bit since those are really just XML with a different schema from the point of view of a library like this.

**Comment by eiriktsarpalis** [comment]
Do you have any examples (e.g. a prototype repo) showing how these interfaces are implemented and how they are meant to be consumed by parsers? If I'm consuming an `IElement`, how can I tell what the right `TElement` type parameter is when I call one of the `TryGetProperty` methods?

**Comment by eerhardt** [comment]
FYI - @agocke. This looks a lot like https://github.com/agocke/serde-dn.

**Comment by agocke** [comment]
Yup, I think this is serde-dn. I would like to revisit putting serde-dn into the framework. After looking at the EventSource serialization implementation, I think that could be solved with serde-dn. I think having a language-agnostic approach to solving reflection-free serialization problems is going to be important for the long-tail of framework serialization issues where we either can't or don't want to build individual source generators for each format.

**Comment by Foxtrek64** [comment]
I can put together a prototype repo, certainly. I was experimenting with an `ISerializationOperators` interface as an expected member, sort of following the example laid out by the new `INumber<TSelf>` and its sub-types.

That said, serde-dn looks like it definitely satisfies the requirements I'm looking for here. I'm perfectly happy with modifying this proposal to instead be the merging of serde-dn into the framework.

**Comment by Foxtrek64** [comment]
Renaming proposal to System.Text.Serialization as I feel the name is more accurate than .Parsing.

**Comment by krwq** [comment]
What's the end goal of this abstraction? We want to create generic DOM for all serializers?

IMO there are nuanced differences between various serializers and various people have different preferences. I think this kind of abstraction should live in the application layer to decide what defaults are appropriate for specific format i.e. for XML when would you prefer attribute over sub-element or reverse?

I'd need to see more evidence of use cases and what's the benefit of using this rather than going direct. Is it really that common for people to switch between different formats? Why not create serialization model and just use serializers, abstraction would be couple of lines of code and faster than using DOM APIs?

**Comment by eiriktsarpalis** [comment]
There are benefits to creating format-agnostic serializers and other platforms have seen success in using these: serde in Rust, pickling in Scala and FsPickler in F# are few examples. It means that you have to deal with the immense complexity of writing a serializer once and extending support to a specific format is simply a matter of implementing a couple of IFormatReader/IFormatWriter interfaces.

The are downsides too, most importantly such interfaces typically make implicit assumptions about the grammar that the underlying format supports. So if it's been designed around JSON then it tends to produce awkward/unidiomatic XML.

**Comment by agocke** [comment]
One of the main benefits is that it allows reuse of the source generator, so that you only need one source generator to support many formats. This massively reduces the investment needed to make things AOT compatible.

**Comment by krwq** [comment]
Per offline conversation I'm assigning @agocke for long-term driving this (likely won't happen in 8.0, possibly for 9.0)

**Comment by Ayazcan3333** [comment]
// I have only included public interfaces from JsonDocument here, but it may be prudent to expose some internal methods as well.
public interface ITextDocument<TRoot, TextWriter> : IDisposable
    where TRoot : IElement
    where TextWriter : ITextWriter
{
    void WriteTo(TextWriter writer);
}
